### PR TITLE
dpsgd

### LINKED ISSUE #264: int labels for RMIA
# Issue
## Problem Description
RMIA requires labels to be integers. Currently, an assertion checks this condition. The assert should be removed and replaced with converting labels to int if they are not already. 

## Expected Behavior
Convert float labels to int for RMIA

## What Needs to be Done
if statement should replace the assert

## How Can It Be Tested or Reproduced
The code is running

### LINKED ISSUE #265: dpsgd for LoS
# Feature

## Desired Behavior / Functionality
Evaluating attack performance when training the target model and/or shadow models with Differentially Private Stochastic Gradient Descent (dpsgd)

## What Needs to Be Done
Using Opacus library to train the target model and shadow models with dpsgd

## How Can It Be Tested
Code is running

### LINKED ISSUE #266: hard coded path dpsgd
# Issue
## Problem Description
In LOS/mimic_GRUD_dpsgd_handler.py, the path to the .pkl file storing DPSGD properties is currently hardcoded. For long-term maintainability, this should be removed. If DPSGD is needed for training shadow models, the path should be moved to a more appropriate location. Otherwise, if supporting shadow model training with DPSGD is unnecessary, the line should be removed entirely

## What Needs to be Done
Evaluate whether using DPSGD in shadow model training improves attack performance in more attackable examples (CIFAR or Celeb HQ) and decide what makes the most sense based on the results.

### PR DESCRIPTION
**Pull request by fazelehh**
# Description

Summary of changes

* Adding dpsgd to the LoS example
* Shadow models can train with or without dpsgd

## Resolved Issues
- [x] fixes #265 
- [x] fixes #264  

## How Has This Been Tested?
Example is running 
## Related Pull Requests

- #PR

### CODE REVIEW

File: examples/mia/LOS/mimic_GRUD_dpsgd_handler.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                                            sample_rate = sample_rate ,
+                                            epochs = privacy_engine_dict["epochs"],
+                                            epsilon_tolerance = privacy_engine_dict["epsilon_tolerance"],
+                                            accountant = "prv",
+                                            eps_error = privacy_engine_dict["eps_error"],)
+        except:
+            # the prv accountant is not robust to large epsilon (even epsilon = 10)
+            # so we will use rdp when it fails, so the actual epsilon may be slightly off
+            # see https://github.com/pytorch/opacus/issues/604
+            noise_multiplier = get_noise_multiplier(target_epsilon = 2,
```
Comment (johanos1) [inline examples/mia/LOS/mimic_GRUD_dpsgd_handler.py:66]: why epsilon=2 here?

File: examples/mia/LOS/mimic_GRUD_dpsgd_handler.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                                                    target_delta = privacy_engine_dict["target_delta"],
+                                                    sample_rate = sample_rate,
+                                                    epochs = privacy_engine_dict["epochs"],
+                                                    epsilon_tolerance = privacy_engine_dict["epsilon_tolerance"],
+                                                    accountant = "rdp")
+
+        # make the model private
+        privacy_engine = PrivacyEngine(accountant = "prv")
```
Comment (johanos1) [inline examples/mia/LOS/mimic_GRUD_dpsgd_handler.py:74]: isnt this off if the accountant used before was rdp?

File: examples/mia/LOS/mimic_GRUD_dpsgd_handler.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                optimizer.zero_grad()
+                output = model(x)
+
+                loss = criterion(output.squeeze(), labels.squeeze())
+                loss.backward()
+                optimizer.step()
+                train_loss += loss.item()
+
+                # Collect predictions and labels for last epoch
+                binary_predictions = sigmoid(output).squeeze().round().cpu().detach().numpy()
```
Comment (johanos1) [inline examples/mia/LOS/mimic_GRUD_dpsgd_handler.py:107]: necessary as you use BCEWithLogitsLoss?

File: examples/mia/LOS/mimic_GRUD_dpsgd_handler.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        return optim.Adam(model.parameters(), lr=learning_rate)
+
+    def train(
+        self,
+        dataloader: DataLoader,
+        model: nn.Module = None,
+        criterion: nn.Module = None,
+        optimizer: optim.Optimizer = None,
+        epochs: int = None,
+    ) -> dict:
```
Comment (johanos1) [inline examples/mia/LOS/mimic_GRUD_dpsgd_handler.py:37]: should return TrainingConfig

**Review by johanos1 — CHANGES_REQUESTED** [review]

File: examples/mia/LOS/mimic_GRUD_dpsgd.ipynb [hunk 5]
```diff
@@ -0,0 +1,1347 @@
+{
```
Comment (johanos1) [inline examples/mia/LOS/mimic_GRUD_dpsgd.ipynb:1]: <p>Please clean all outputs.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (johanos1) [inline examples/mia/LOS/mimic_GRUD_dpsgd.ipynb:1]: <p>Explain the params in a cell above (markdown)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (johanos1) [inline examples/mia/LOS/mimic_GRUD_dpsgd.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    # epsilons = [.0001, .001, .01, .1, .5, 1, 2, 3.5, 5, 7, 10] # epsilons to run over
</pre><p>?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

File: .gitignore
[the remark this answers is not in the document]
Comment (fazelehh) [reply .gitignore:1]: We've talked about this quite a few times, and it seems like no final decision was made, or at least I don't remember it.
My take: Keeping the 'data' directory—especially in examples where we ask users to place their data there, like LoS—makes it clearer where the folder should be. We just need to remember to add it in gitignore (which is the case here; we have a .gotignore in LoS ignoring the data)

I remember you reasoned that we could simply run mkdir data.

[the remark this answers is not in the document]
Comment (fazelehh) [reply .gitignore:1]: I think I can go with your preference if my approach feels too risky! 😊 So, I'll add data/ to .gitignore and update the prepare-data instructions to include a mkdir data step

File: examples/mia/LOS/mimic_GRUD_dpsgd_handler.py (hunk 3 above)
Comment (fazelehh) [reply examples/mia/LOS/mimic_GRUD_dpsgd_handler.py:107]: Actually, I think we should use sigmoid, and we should also revert adult_handler.py to its previous version. 

I believe the confusion came from how we handle the model's output. The criterion (BCEWithLogitsLoss) already includes sigmoid, so we don’t need to apply it before computing the loss. However, when making predictions, we’re using the raw model output, which hasn’t been affected by the sigmoid in the criterion. That’s why we do need to apply sigmoid to get the final predictions.

**Review by johanos1 — APPROVED** [review]
Looks good!
